### DISCUSSION TITLE
avoiding typescript error `@typescript-eslint/no-unsafe-assignment` in `objectContaining()`

### DISCUSSION THREAD

**Discussion by kevinlaw91** [vitest-dev/vitest#6614] [Q&A] [answered]
`expect.any()` returns type `any`. How do you make typescript happy?

```ts
expect(newAccount)
  .toEqual(
    expect.objectContaining({
      id: expect.any(String), // Unsafe assignment of an `any` value.
    }),
  );
```

I'm currently using `as` operator, not sure if this is the correct solution or there is a better way
```ts
id: expect.any(String) as string
```

**Comment by kevinlaw91** [comment]
Disable eslint rule as suggested here:
https://github.com/vitest-dev/vitest/issues/4543#issuecomment-1824628142

**Comment by CyberT33N** [comment, marked as answer]
# ESLint-Compliant Matcher Strategy for `toMatchObject()` in Vitest

## ❗ Critical Rules

* **MUST:** When using asymmetric matchers inside `toMatchObject()`, you **MUST** apply one of the prioritized strategies to avoid ESLint error `@typescript-eslint/no-unsafe-assignment`.
* **MUST:** The **preferred solution** is to use custom, type-safe matcher utilities instead of `expect.any(Type)`.
* **MUST NOT:** You **MUST NOT** use `expect.any(Type)` directly in `toMatchObject()` when ESLint with `no-unsafe-assignment` is enabled.
* **CONSIDER:** The order of preferred approaches is: 1. Custom matcher utilities, 2. Separate assertions, 3. Typecasts, 4. ESLint-disable as a last resort.

## The Problem

When using `expect.any(Type)` inside `toMatchObject()` in **Vitest**, **ESLint** with `@typescript-eslint/no-unsafe-assignment` throws the following error:

> `"Unsafe assignment of an any value."`

This happens because `expect.any(...)` internally returns a value of type `any`, which clashes with strict TypeScript settings.

## ✅ Examples

### 🥇 Preferred: Custom Matcher Utility

<example>
```typescript
// test-utils/AsymmetricMatchers.ts
interface IMatcher {
  asymmetricMatch(received: unknown): boolean
  toString(): string
  getExpectedType(): string
}

function makeMatcher(name: string, type: string, matchFn: (value: unknown) => boolean): IMatcher {
return {
asymmetricMatch: matchFn,
toString: () => name,
getExpectedType: () => type
}
}

export const AnyDate = makeMatcher('AnyDate', 'Date', value =>
value instanceof Date && !isNaN(value.getTime())
)

export const AnyString = makeMatcher('AnyString', 'String', value =>
typeof value === 'string'
)

export const AnyNumber = makeMatcher('AnyNumber', 'Number', value =>
typeof value === 'number' && !isNaN(value)
)

export const AnyBoolean = makeMatcher('AnyBoolean', 'Boolean', value =>
typeof value === 'boolean'
)

export const AnyObject = makeMatcher('AnyObject', 'Object', value =>
typeof value === 'object' && value !== null && !Array.isArray(value)
)

export const AnyArray = makeMatcher('AnyArray', 'Array', value =>
Array.isArray(value)
)

export const Anything = makeMatcher('Anything', 'Defined', value =>
value !== null && value !== undefined
)

export const AnyFunction = makeMatcher('AnyFunction', 'Function', value =>
typeof value === 'function'
)

type MatcherName =
\| 'Date' | 'String' | 'Number' | 'Boolean' | 'Object' | 'Array' | 'Function' | 'Defined'

export const Any = (type: MatcherName): IMatcher => {
const map: Record\<MatcherName, IMatcher> = {
Date: AnyDate,
String: AnyString,
Number: AnyNumber,
Boolean: AnyBoolean,
Object: AnyObject,
Array: AnyArray,
Function: AnyFunction,
Defined: Anything
}
return map\[type]
}
````
</example>

<example>
```typescript
// ✅ CORRECT: Using custom matcher utilities
import { Any } from 'test-utils/AsymmetricMatchers'

test('returns finding with createdAt', () => {
  const result = {
    pvs01Finding: {
      createdAt: new Date(),
      name: 'Test'
    }
  }

  expect(result).toMatchObject({
    pvs01Finding: {
      createdAt: Any('Date'),
      name: Any('String')
    }
  })
})
````

</example>

### 🥈 Alternative: Separate Assertions

<example>
```ts
// ✅ CORRECT: Type assertions outside the match object
test('should validate types separately', () => {
  const result = {
    pvs01Finding: {
      createdAt: new Date(),
      name: 'Test'
    }
  }

// Structure assertion
expect(result).toMatchObject({
pvs01Finding: {
name: 'Test'
}
})

// Separate type assertion
expect(result.pvs01Finding.createdAt).toEqual(expect.any(Date))
})

````
</example>

### 🥉 Workaround: Typecast

<example>
```ts
// ⚠️ ACCEPTABLE: Typecast workaround
expect(result).toMatchObject({
  pvs01Finding: {
    createdAt: expect.any(Date as unknown as typeof Date)
  }
})
````

</example>

### 🟥 Last Resort: ESLint Disable

<example>
```ts
// 🚨 LAST RESORT: Disable ESLint rule
expect(result).toMatchObject({
  pvs01Finding: {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    createdAt: expect.any(Date)
  }
})
```
</example>

### ❌ Incorrect Usage

<example type="invalid">
```ts
// ❌ WRONG: Direct expect.any() inside toMatchObject()
expect(result).toMatchObject({
  pvs01Finding: {
    createdAt: expect.any(Date), // VIOLATION: Unsafe assignment of an any value
    name: expect.any(String)     // VIOLATION: Unsafe assignment of an any value
  }
})
```
</example>

## 📊 Solution Comparison

| Priority | Solution                       | Inline Use | Type-Safe | Linter-Safe | Readable | Autocomplete Support |
| -------: | ------------------------------ | ---------- | --------- | ----------- | -------- | -------------------- |
|    🥇 1. | Custom `Any(...)` Matchers     | ✅          | ✅         | ✅           | ✅        | ✅                    |
|    🥈 2. | Separate `expect(...).toEqual` | ❌          | ✅         | ✅           | ✅        | ✅                    |
|    🥉 3. | Typecast                       | ✅          | ❌         | ⚠️          | ❌        | ❌                    |
|    🟥 4. | ESLint Disable Comment         | ✅          | ✅         | ❌           | ❌        | ✅                    |

## 🤔 Rationale / Why This Rule Matters

* **ESLint Compliance:** Avoids `@typescript-eslint/no-unsafe-assignment` errors in strict TypeScript projects.
* **Type Safety:** Custom matcher utilities provide better type safety than `expect.any()`.
* **Maintainability:** A consistent matcher strategy improves test clarity and extensibility.
* **Readability:** Descriptive matcher names like `Any('Date')` are more informative than `expect.any(Date)`.
* **Extensibility:** Custom matchers can be extended to support project-specific types (e.g., RegExp, Set, Map).
* **Future-Proof:** Remains unaffected by internal changes to `expect.any()` in Vitest.

**Reply by kevinlaw91** [reply]
interesting. where did you find this? is it from chatgpt?
